src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/RefDictionary.cs

```diff
@@ -0,0 +1,180 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Experimental.Collections
+{
+    public class RefDictionary<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey> // almost IReadOnlyDictionary<TKey, TValue>
+    {
+        private int[] _buckets;
+        private Entry[] _entries;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            public int next;
+        }
+
+        public RefDictionary(int capacity = 16)
+        {
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetBucketIndex(TKey key) => (key.GetHashCode() & 0x7FFFFFFF) % _buckets.Length;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(TKey key) => _buckets[GetBucketIndex(key)] - 1;
+
+        public int Count { get; private set; }
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return true;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return entries[entryIndex].value;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return default;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                int entryIndex = GetEntryIndex(key);
+
+                while (entryIndex != -1)
+                {
+                    if (entries[entryIndex].key.Equals(key))
+                    {
+                        return ref entries[entryIndex].value;
+                    }
+                    entryIndex = entries[entryIndex].next;
+                }
+
+                if (Count == entries.Length) entries = Resize();
+
+                entryIndex = Count++;
+                entries[entryIndex].key = key;
+                int bucket = GetBucketIndex(key);
+                entries[entryIndex].next = _buckets[bucket] - 1;
+                _buckets[bucket] = entryIndex + 1;
+                return ref entries[entryIndex].value;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Are you really getting any meaningful performance win by AggressiveInlining `Resize()` method? Resize is called rarely - AggressiveInlining it into Add will likely slow down the more frequent Add operation.

Reply:
Nothing that can be resolved with BenchmarkDotNet. Have taken it out.